File: apps/dsa-web/src/types/market.ts
/** Market workbench contracts returned by the chart/annotation API. */

export interface MarketCandle {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number | null;
  amount?: number | null;
  changePercent?: number | null;
  indicators: {
    ma5?: number | null;
    ma10?: number | null;
    ma20?: number | null;
    rsi14?: number | null;
    macd?: number | null;
    macdSignal?: number | null;
    macdHistogram?: number | null;
  };
}

export interface MarketCandleResponse {
  securityId: Record<string, unknown>;
  period: string;
  candles: MarketCandle[];
  source: string;
  sourceStatus: string;
  fallbackChain: string[];
  asOf?: string | null;
  delaySeconds?: number | null;
  stale: boolean;
  dataQuality: string;
  limitations: string[];
}

export interface MarketAnnotation {
  type: string;
  source: string;
  timestamp?: string | null;
  accountId?: number | null;
  symbol: string;
  payload: Record<string, unknown>;
}

export interface MarketAnnotationResponse {
  symbol: string;
  start?: string | null;
  end?: string | null;
  items: MarketAnnotation[];
  page: number;
  pageSize: number;
  total: number;
  hasMore: boolean;
  partial: boolean;
  limitations: string[];
}

File: api/v1/schemas/market.py
# -*- coding: utf-8 -*-
"""Market chart and annotation response schemas."""

from __future__ import annotations

from datetime import date
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class CandleItem(BaseModel):
    timestamp: str
    open: float
    high: float
    low: float
    close: float
    volume: Optional[float] = None
    amount: Optional[float] = None
    change_percent: Optional[float] = None
    indicators: Dict[str, Optional[float]] = Field(default_factory=dict)


class CandleResponse(BaseModel):
    security_id: Dict[str, Any]
    period: str
    candles: List[CandleItem] = Field(default_factory=list)
    source: str
    source_status: str
    fallback_chain: List[str] = Field(default_factory=list)
    as_of: Optional[str] = None
    delay_seconds: Optional[float] = None
    stale: bool = False
    data_quality: str
    limitations: List[str] = Field(default_factory=list)


class MarketSnapshotResponse(BaseModel):
    security_id: Dict[str, Any]
    quote: Optional[CandleItem] = None
    source: str
    source_status: str
    as_of: Optional[str] = None
    delay_seconds: Optional[float] = None
    stale: bool = False
    data_quality: str
    limitations: List[str] = Field(default_factory=list)


class AnnotationItem(BaseModel):
    type: str
    source: str
    timestamp: Optional[str] = None
    account_id: Optional[int] = None
    symbol: str
    payload: Dict[str, Any] = Field(default_factory=dict)


class AnnotationResponse(BaseModel):
    symbol: str
    start: Optional[str] = None
    end: Optional[str] = None
    items: List[AnnotationItem] = Field(default_factory=list)
    page: int = 1
    page_size: int = 100
    total: int = 0
    has_more: bool = False
    partial: bool = False
    limitations: List[str] = Field(default_factory=list)
